src/SessionSetup/ConnectToTargetDialog.cpp

```diff
@@ -31,160 +32,143 @@ ConnectToTargetDialog::ConnectToTargetDialog(
-              [this](ErrorMessageOr<orbit_ggp::SshInfo> ssh_info_result) {
-                OnSshInfoAvailable(std::move(ssh_info_result));
-              });
 
-  return true;
-}
+  auto process_future = ggp_client_->GetSshInfoAsync(instance_id_, std::nullopt);
+  auto instance_future = ggp_client_->DescribeInstanceAsync(instance_id_);
+  auto joined_future = orbit_base::JoinFutures(process_future, instance_future);
 
-void ConnectToTargetDialog::OnSshInfoAvailable(ErrorMessageOr<orbit_ggp::SshInfo> ssh_info_result) {
-  if (!ssh_info_result.has_error()) {
-    LOG("Received ssh info for instance with id: %s", instance_id_.toStdString());
+  std::optional<TargetConfiguration> target;
 
-    auto ssh_info = ssh_info_result.value();
-    auto deployment_result = DeployOrbitService(ssh_info);
-    if (deployment_result.has_value()) {
-      result_data_.grpc_channel_ = CreateGrpcChannel(deployment_result.value().grpc_port);
-      ListProcessesAndSaveProcess(result_data_.grpc_channel_);
-    }
-  } else {
-    result_data_.error_message_ =
-        QString("Error receiving ssh info for instance with id %1:%2")
-            .arg(instance_id_, QString::fromStdString(ssh_info_result.error().message()));
+  joined_future.Then(main_thread_executor_.get(),
+                     [this, &target](MaybeSshAndInstanceData ssh_instance_data) {
+                       auto result = OnAsyncDataAvailable(std::move(ssh_instance_data));
+                       if (result.has_value()) {
+                         target = std::move(result.value());
+                         accept();
+                       } else {
+                         LogAndDisplayError(QString::fromStdString(result.error().message()));
+                         reject();
+                       }
+                     });
+
+  int rc = QDialog::exec();
+
+  if (rc != QDialog::Accepted) {
+    return std::nullopt;
   }
 
-  CloseDialogIfResultIsReady();
+  return target;
 }
 
-void ConnectToTargetDialog::LoadInstanceDetailsAsync() {
-  ggp_client_->DescribeInstanceAsync(instance_id_)
-      .Then(main_thread_executor_.get(),
-            [this](ErrorMessageOr<orbit_ggp::Instance> instance_result) {
-              if (instance_result.has_error()) {
-                result_data_.maybe_instance_ = std::nullopt;
-                result_data_.error_message_ =
-                    QString::fromStdString(instance_result.error().message());
-              } else {
-                result_data_.maybe_instance_ = instance_result.value();
-              }
-              CloseDialogIfResultIsReady();
-            });
-}
+ErrorMessageOr<StadiaTarget> ConnectToTargetDialog::OnAsyncDataAvailable(
+    MaybeSshAndInstanceData ssh_instance_data) {
+  auto& [ssh_info_result, instance_result] = ssh_instance_data;
+
+  std::string error;
+  if (ssh_info_result.has_error()) {
+    return ssh_info_result.error();
+  } else if (instance_result.has_error()) {
+    return instance_result.error();
+  }
```

Comment:
pers pref:
`OUTCOME_TRY(auto&& ssh_info, ssh_info_result);`
Or even `OUTCOME_TRY(auto&& ssh_info, ssh_instance_data.first);`

(and for instance as well)